core/vm/contracts.go

```diff
@@ -1115,51 +1114,56 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
 	return g.EncodePoint(r), nil
 }
 
-var errVoteSignatureVerify = errors.New("invalid signatures")
-
-// voteSignatureVerify implements BEP-126 finality signature verification precompile.
-type voteSignatureVerify struct{}
+// blsSignatureVerify implements bls signature verification precompile.
+type blsSignatureVerify struct{}
 
 // RequiredGas returns the gas required to execute the pre-compiled contract.
-func (c *voteSignatureVerify) RequiredGas(input []byte) uint64 {
-	return params.VoteSignatureVerifyGas
+func (c *blsSignatureVerify) RequiredGas(input []byte) uint64 {
+	return params.BlsSignatureVerifyGas
 }
 
-func (c *voteSignatureVerify) Run(input []byte) ([]byte, error) {
-	var (
-		srcNum  = new(big.Int).SetBytes(getData(input, 0, 32)).Uint64()
-		tarNum  = new(big.Int).SetBytes(getData(input, 32, 32)).Uint64()
-		srcHash = getData(input, 64, 32)
-		tarHash = getData(input, 96, 32)
-		sig     = getData(input, 128, 96)
-		BLSKey  = getData(input, 224, 48)
-	)
-
-	sigs := make([][]byte, 1)
-	msgs := make([][32]byte, 1)
-	pubKeys := make([]bls.PublicKey, 1)
+// Run input:
+// msg      | signature | [{bls pubkey}] |
+// 32 bytes | 96 bytes  | [{48 bytes}]   |
+func (c *blsSignatureVerify) Run(input []byte) ([]byte, error) {
+	minimumLength := uint64(32) + uint64(96)
+	singlePubkeyLength := uint64(48)
 
-	voteData := &types.VoteData{
-		SourceNumber: srcNum,
-		SourceHash:   common.BytesToHash(srcHash),
-		TargetNumber: tarNum,
-		TargetHash:   common.BytesToHash(tarHash),
+	inputLen := uint64(len(input))
+	if inputLen <= minimumLength || (inputLen-minimumLength)%singlePubkeyLength != 0 {
+		return nil, fmt.Errorf("expected input size %d+%d*N, actual input size: %d", minimumLength, singlePubkeyLength, inputLen)
 	}
-	copy(msgs[0][:], voteData.Hash().Bytes())
 
-	pubKey, err := bls.PublicKeyFromBytes(BLSKey)
+	var msg [32]byte
+	msgBytes := getData(input, 0, 32)
+	copy(msg[:], msgBytes)
+
+	signatureBytes := getData(input, 32, 96)
+	sig, err := bls.SignatureFromBytes(signatureBytes)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("invalid signature: %v", err)
 	}
-	pubKeys[0] = pubKey
-	sigs[0] = sig
 
-	success, err := bls.VerifyMultipleSignatures(sigs, msgs, pubKeys)
-	if err != nil {
-		return nil, err
+	pubKeyNumber := (inputLen - minimumLength) / singlePubkeyLength
+	pubKeys := make([]bls.PublicKey, pubKeyNumber)
+	for i := uint64(0); i < pubKeyNumber; i++ {
+		pubKeyBytes := getData(input, 128+i*singlePubkeyLength, 48)
+		pubKey, err := bls.PublicKeyFromBytes(pubKeyBytes)
+		if err != nil {
+			return nil, err
+		}
+		pubKeys[i] = pubKey
 	}
-	if !success {
-		return nil, errVoteSignatureVerify
+
+	if pubKeyNumber > 1 {
+		if !sig.FastAggregateVerify(pubKeys, msg) {
+			return nil, fmt.Errorf("signature verify failed")
+		}
+	} else {
+		if !sig.Verify(pubKeys[0], msgBytes) {
```

Comment:
It will panic here if the len of `pubKeys` is zero, it is possible for current implementation.